packages/expo-image/src/useImage.ts

```diff
@@ -0,0 +1,48 @@
+'use client';
+
+import { useEffect, useState } from 'react';
+
+import { Image } from './Image';
+import type { ImageRef, ImageSource, UseImageHookOptions } from './Image.types';
+import { resolveSource } from './utils/resolveSources';
+
+/**
+ * A hook that loads an image from the given source and returns a reference
+ * to the native image instance, or `null` until the image is successfully loaded.
+ * @platform android
+ * @platform ios
+ */
+export function useImage(
+  source: ImageSource | string,
+  options: UseImageHookOptions = {}
+): ImageRef | null {
+  const resolvedSource = resolveSource(source) as ImageSource;
+  const [image, setImage] = useState<ImageRef | null>(null);
+
+  useEffect(() => {
+    async function loadImage() {
+      try {
+        const image = await Image.loadAsync(resolvedSource);
+        setImage(image);
+      } catch (error) {
+        if (options.onError) {
+          options.onError(error, loadImage);
+        } else {
+          // Print unhandled errors to the console.
+          console.error(
+            `Loading an image from '${resolvedSource.uri}' failed, use 'onError' option to handle errors and suppress this message`
+          );
+          console.error(error);
+        }
+      }
+    }
+
+    loadImage();
+
+    return () => {
+      image?.release();
```

Comment:
I wonder what happens given:

```
const [sourceUri, setSourceUri] = useState<string>(getRandomImageUri());
const image = useImage(sourceUri);

useEffect(() => {
// call setSourceUri with a random image after 2 seconds 
},[])

return <Image style={styles.image} source={image} />
```

1.  `<Image />` is rendering the 1st image
2. the `sourceUri` changes. `release()` is called on the 1st image, but `<Image />` still is rendering the 1st image until the second is loaded. What happens between the `release` and the 2nd load?

Reply:
The image will still render the first image, even after releasing it. `release` only detaches the JS and native object from themselves to give the native side full control on when the native object is deallocated (by automatic reference counting rather than GC in the JS engine).
In this particular case, the underlying `UIImage` is assigned to `UIImageView` which keeps a strong reference to that image. So "releasing" doesn't have to necessarily destroy it, it's up to ARC now.

Reply:
TY, good to know!